### ISSUE TITLE
Adding differences and moving average features

### ISSUE THREAD

**Issue by msat59** [JoaquinAmatRodrigo/skforecast#499]
Hi guys.

Has anyone tried to add other ARIMA components to the model? Lags are currently supported which is equal to `AR(p)` model, however, `q` and `d` are missing. I know that these can be created and fed into the model via `exog` variable, but they can be generated automatically by the model such as `ForecasterAutoreg` too.

**Comment by JavierEscobarOrtiz** [comment]
Hello @msat59,

We have two options here:

+ You can use a `ForecasterAutoregCustom` to add additional features to your training matrix, such as the moving average. However, this method doesn't allow differentiation.

Check: https://skforecast.org/latest/user_guides/custom-predictors.html

+ You can use a `ForecasterSarimax` to use all the features of skforecast with an ARIMA model.

Check: https://skforecast.org/latest/user_guides/forecasting-sarimax-arima.html

In skforecast 0.10.0 we will work on adding the ability to differentiate your series for those forecasters that use a scikit-learn regressor. Glad to hear this is something the community is asking for. 😄

**Comment by msat59** [comment]
Hi @JavierEscobarOrtiz.

Thanks for your prompt response. Actually, I am not interested in SARIMAX model as it has a lot of hyperparameters to tune, of which the `trend` is the worst one, possibly in my problems. 
Any interest to add diff and MA parameters to models?

**Comment by JavierEscobarOrtiz** [comment]
MA in a non-ARIMA model is currently possible with the `ForecasterAutoregCustom`, here is an example of how to implement a moving average of the last 20 observations + 10 lags as predictors with a `RandomForestRegressor`:

(more detailed here: https://skforecast.org/latest/user_guides/custom-predictors.html)

```python
# Libraries
# ==============================================================================
import numpy as np
import pandas as pd
from skforecast.ForecasterAutoregCustom import ForecasterAutoregCustom
from sklearn.ensemble import RandomForestRegressor

# Download data
# ==============================================================================
url = (
    'https://raw.githubusercontent.com/JoaquinAmatRodrigo/skforecast/master/' 
    'data/h2o.csv'
)
data = pd.read_csv(url, sep=',', header=0, names=['y', 'datetime'])

# Data preprocessing
# ==============================================================================
data['datetime'] = pd.to_datetime(data['datetime'], format='%Y-%m-%d')
data = data.set_index('datetime')
data = data.asfreq('MS')
data = data['y']
data = data.sort_index()

# Custom function to create predictors
# ==============================================================================
def create_predictors(y):
    """
    Create first 10 lags of a time series.
    Calculate moving average with window 20.
    """

    lags = y[-1:-11:-1]     # window_size = 10
    mean = np.mean(y[-20:]) # window_size = 20
    predictors = np.hstack([lags, mean])

    return predictors

# Create and fit forecaster
# ==============================================================================
forecaster = ForecasterAutoregCustom(
                 regressor      = RandomForestRegressor(random_state=123),
                 fun_predictors = create_predictors,
                 name_predictors = [f'lag {i}' for i in range(1, 11)] + ['moving_avg_20'],
                 window_size    = 20 # window_size needed by the mean is the most restrictive one
             )

forecaster.fit(y=data)
forecaster
```

As I said, in skforecast 0.10.0 we will be working on adding the ability to differentiate your series for those forecasters using a scikit-learn regressor. If you have any suggestions, we are more than happy to hear from you.

**Comment by JoaquinAmatRodrigo** [comment]
Hi @msat59 
I wanted to add some information to what @JavierEscobarOrtiz explained earlier. When modeling a time series with a trend, it is advisable to consider using a linear regressor like Ridge or Lasso. These models have the ability to extrapolate, which can be useful in capturing the trend accurately. On the other hand, tree-based models such as random forest and gradient boosting are not able.

**Comment by msat59** [comment]
Hi @JoaquinAmatRodrigo.

That was really good advice; I never read anywhere about the logic behind using linear regression in time-series forecasting (what's being used in ARIMA).

**Comment by edgBR** [comment]
Hi,

Scikit-learn fashion pipeline:
```python

def rolling_average_transformer(X: Union[pd.Series, pd.DataFrame], window: int = 5):
    """Compute the moving average with a length of 'window'."""

    if isinstance(X, pd.Series):
        X = X.to_frame()

    X_average = X.rolling(window=window).mean()
    X_average.columns = [f"{col}_rolling_average_{window}" for col in X.columns]

    logger.debug(f"rolling_average_transformer (transform): {X_average.shape}")

    return X_average


def rolling_std_transformer(X: Union[pd.Series, pd.DataFrame], window: int = 5):
    """Compute the moving standard deviation with a length of 'window'."""

    if isinstance(X, pd.Series):
        X = X.to_frame()

    X_std = X.rolling(window=window).std()
    X_std.columns = [f"{col}_rolling_std_{window}" for col in X.columns]

    logger.debug(f"rolling_std_transformer (transform): {X_std.shape}")

    return X_std


def rolling_median_transformer(X: Union[pd.Series, pd.DataFrame], window: int = 5):
    """Compute the moving median with a length of 'window'."""

    if isinstance(X, pd.Series):
        X = X.to_frame()

    X_median = X.rolling(window=window).median()
    X_median.columns = [f"{col}_rolling_median_{window}" for col in X.columns]

    logger.debug(f"rolling_median_transformer (transform): {X_median.shape}")

    return X_median
```

Then:
```python
RollingAverageTransformer = FunctionTransformer(partial(rolling_average_transformer, window=forecast_horizon))
```

**Comment by JavierEscobarOrtiz** [comment]
Hello,

Differentiation is now possible with skforecast 0.10.0, check:

https://skforecast.org/latest/faq/time-series-differentiation

and

[Modelling time series trend with tree based models](https://www.cienciadedatos.net/documentos/py49-modelling-time-series-trend-with-tree-based-models.html)

Hopt it helps!
